README.md
# Car Share Back

[![Build Status](https://travis-ci.org/LewisWatson/carshare-back.svg?branch=master)](https://travis-ci.org/LewisWatson/carshare-back)
[![Coverage Status](https://coveralls.io/repos/github/LewisWatson/carshare-back/badge.svg?branch=feature%2Ffirebase-jwt-auth)](https://coveralls.io/github/LewisWatson/carshare-back)
[![GitPitch](https://gitpitch.com/assets/badge.svg)](https://gitpitch.com/LewisWatson/carshare-ninja-pitch/master?grs=github&t=white)
[![stability-experimental](https://img.shields.io/badge/stability-experimental-orange.svg)](https://github.com/emersion/stability-badges#experimental)

An Open Source API for tracking car shares.

Designed to enable easy tracking of the distance members travel as passengers and as drivers. Car share members can make informed day to day decisions about who should drive next based on the ratio of distance travelled as the driver vs as a passenger.

Written in [Go] and designed using [{json:api}] specification for building API's in JSON.

## Install

```bash
go get github.com/LewisWatson/carshare-back
```

## Run

```bash
docker run -d -p 27017:27017 mongo --smallfiles
$GOPATH/bin/carshare-back
1970/01/01 00:00:00 connecting to mongodb server via url: localhost
1970/01/01 00:00:00 listening on :31415
```

### Configuration

```bash
$GOPATH/bin/carshare-back --help
usage: carshare-back [<flags>]

API for tracking car shares

Flags:
  --help                        Show context-sensitive help (also try --help-long and --help-man).
  --port=31415                  Set port to bind to
  --mgoURL=localhost            URL to MongoDB server or seed server(s) for clusters
  --firebase="ridesharelogger"  Firebase project to use for authentication
  --cors=URI                    Enable HTTP Access Control (CORS) for the specified URI
  --version                     Show application version.
```

## Docker

The [Dockerfile](Dockerfile) uses a [minimal Docker image based on Alpine Linux](https://hub.docker.com/_/alpine/) with a different implimentation of libc. Therefore, it is important that a static binary is used when building

```bash
go build -tags 'gingonic netgo' -ldflags '-extldflags "-lm -lstdc++ -static"'
docker build .
```

Pre-made images are available as [lewiswatson/carshare-back](https://hub.docker.com/r/lewiswatson/carshare-back/)

## Available endpoints

| OPTIONS | GET | POST | PATCH | DELETE | Path
| --------|-----|------|-------|--------| -------------------------------------------
| OPTIONS |     | POST |       |        | /v0/users
| OPTIONS |     |      | PATCH | DELETE | /v0/users/:id
| OPTIONS |     | POST |       |        | /v0/trips
| OPTIONS | GET |      |       |        | /v0/trips/:id
|         | GET |      | PATCH |        | /v0/trips/:id/relationships/carShare
|         | GET |      |       |        | /v0/trips/:id/carShare
|         | GET |      | PATCH |        | /v0/trips/:id/relationships/driver
|         | GET |      |       |        | /v0/trips/:id/driver
|         | GET | POST | PATCH | DELETE | /v0/trips/:id/relationships/passengers
|         | GET |      |       |        | /v0/trips/:id/passengers
|         |     |      | PATCH | DELETE | /v0/trips/:id
| OPTIONS | GET | POST |       |        | /v0/carShares
| OPTIONS | GET |      | PATCH | DELETE | /v0/carShares/:id
|         | GET | POST | PATCH | DELETE | /v0/carShares/:id/relationships/trips
|         | GET |      |       |        | /v0/carShares/:id/trips
|         | GET | POST | PATCH | DELETE | /v0/carShares/:id/relationships/members
|         | GET | POST | PATCH | DELETE | /v0/carShares/:id/members
|         | GET |      |       |        | /v0/carShares/:id/relationships/admins
|         | GET |      |       |        | /v0/carShares/:id/admins
|         | GET |      |       |        | /metrics

### Metrics

The `/metrics` endpoint exposes internal metrics for [prometheus](https://prometheus.io/) monitoring.

## License

Copyright 2017 Lewis Watson

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

[mongoDB]: https://www.mongodb.com/
[{json:api}]: http://jsonapi.org
[Go]: https://golang.org/


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [0.5.0] - 2017-11-14

### Changed

- Auto-onboarding of valid Firebase users. 
- Introduce dependency management with golang/dep after run in with dependecncy hell
- Target Go 1.9

## [0.4.1] - 2017-05-03

### Fixed

- Ensure the database connection is always stored in the context, not just
  when CORS is enabled.

## [0.4.0] - 2017-04-04

### Added

- Prometheus metrics collection via `/metrics` endpoint

### Changed

- Add access control to trip resource
- New trips need to have a valid car share on creation
- Switch to Gin Framework
- Switch to https://github.com/op/go-logging for more flexible logging
- Make firebase project configurable

## [0.3.3] - 2017-03-27

### Fixed

- Use full mgoURL from command line to connect to database

## [0.3.2] - 2017-03-23

### Changed

- Switch base docker image to alpine, reduces image size from 200mb+ to less than 20mb
- Command line arguments working via docker run
  ```bash
  docker run carshare-back --port 1337 --mgoURL mongo
  2017/03/23 07:25:55 connecting to mongodb server via url: mongo
  2017/03/23 07:25:56 listening on :1337
  ```

## [0.3.1] - 2017-03-21

### Fixed

- Re-introduce environmental variable configuration support (for docker)

## [0.3.0] - 2017-03-21

### Added

- CORS headers
- Firebase JWT Validation
- Restrict access by user

### Changed

- Update to Go 1.8
- Command line configuration instead of environment variables

### Fixed

- Relation links between car shares, trips and users

## [0.2.0] - 2016-12-10

### Added

- MongoDB data store support
- Configure MongoDB URL via `CARSHARE_MGO_URL` environment variable
- Configure server port via `CARSHARE_PORT` environment variable
- Created Dockerfile

### Changed

- Standardisd ID's on [BSON ObjectId](https://docs.mongodb.com/manual/reference/bson-types/#objectid)
- Overhaul error handling
- Unit test now run twice. First with the fast in-memory data store (fail fast), then as an integration test against a MongoDB docker container


## [0.1.0] - 2016-11-12

### Added

- Create basic functionality with in memory data store
- Add ability to create users, car shares and trips via json:api REST interface and store in simple in memory data store
- Add README and CHANGELOG

[Unreleased]:https://github.com/LewisWatson/carshare-back/compare/v0.4.1...HEAD
[0.4.1]: https://github.com/LewisWatson/carshare-back/releases/tag/v0.4.1
[0.4.0]: https://github.com/LewisWatson/carshare-back/releases/tag/v0.4.0
[0.3.3]: https://github.com/LewisWatson/carshare-back/releases/tag/v0.3.3
[0.3.2]: https://github.com/LewisWatson/carshare-back/releases/tag/v0.3.2
[0.3.1]: https://github.com/LewisWatson/carshare-back/releases/tag/v0.3.1
[0.3.0]: https://github.com/LewisWatson/carshare-back/releases/tag/v0.3.0
[0.2.0]: https://github.com/LewisWatson/carshare-back/releases/tag/v0.2.0
[0.1.0]: https://github.com/LewisWatson/carshare-back/releases/tag/v0.1.0


carshare-back.go
/*
RESTful API for the car share system
*/
package main

import (
	"fmt"
	"net/http"
	"os"

	"github.com/manyminds/api2go"
	"github.com/manyminds/api2go/routing"
	"github.com/LewisWatson/carshare-back/model"
	"github.com/LewisWatson/carshare-back/resource"
	"github.com/LewisWatson/carshare-back/storage/mongodb"
	"github.com/LewisWatson/firebase-jwt-auth"
	"github.com/alecthomas/kingpin"
	"github.com/benbjohnson/clock"
	"github.com/gin-gonic/gin"
	"github.com/op/go-logging"
	"github.com/prometheus/client_golang/prometheus/promhttp"

	"gopkg.in/mgo.v2"
)

var (
	port              = kingpin.Flag("port", "Set port to bind to").Default("31415").Envar("CARSHARE_PORT").Int()
	mgoURL            = kingpin.Flag("mgoURL", "URL to MongoDB server or seed server(s) for clusters").Default("localhost").Envar("CARSHARE_MGO_URL").URL()
	firebaseProjectID = kingpin.Flag("firebase", "Firebase project to use for authentication").Default("ridesharelogger").Envar("CARSHARE_FIREBASE_PROJECT").String()
	acao              = kingpin.Flag("cors", "Enable HTTP Access Control (CORS) for the specified URI").PlaceHolder("URI").Envar("CARSHARE_CORS_URI").String()

	log    = logging.MustGetLogger("main")
	format = logging.MustStringFormatter(
		`%{color}%{time:2006-01-02T15:04:05.999} %{level:.4s} %{id:03x}%{color:reset} %{message}`,
	)

	userStorage     = &mongodb.UserStorage{}
	carShareStorage = &mongodb.CarShareStorage{}
	tripStorage     = &mongodb.TripStorage{}
)

func init() {

	logging.SetBackend(logging.NewBackendFormatter(logging.NewLogBackend(os.Stderr, "", 0), format))

	kingpin.UsageTemplate(kingpin.CompactUsageTemplate).Version("0.4.0").Author("Lewis Watson")
	kingpin.CommandLine.Help = "API for tracking car shares"
	kingpin.Parse()
}

func main() {

	log.Infof("connecting to mongodb server %s%s", (*mgoURL).Host, (*mgoURL).Path)
	db, err := mgo.Dial((*mgoURL).String())
	if err != nil {
		log.Fatalf("error connecting to mongodb server: %s", err)
	}

	log.Infof("using firebase project \"%s\" for authentication", *firebaseProjectID)
	tokenVerifier, err := fireauth.New(*firebaseProjectID)
	if err != nil {
		log.Fatal(err)
	}

	r := gin.Default()
	api := api2go.NewAPIWithRouting(
		"v0",
		api2go.NewStaticResolver("/"),
		routing.Gin(r),
	)

	if *acao != "" {
		log.Infof("enabling CORS access for %s", *acao)
	}

	api.UseMiddleware(
		func(c api2go.APIContexter, w http.ResponseWriter, r *http.Request) {
			// ensure the db connection is always available in the context
			c.Set("db", db)
			if *acao != "" {
				w.Header().Set("Access-Control-Allow-Origin", *acao)
				w.Header().Set("Access-Control-Allow-Headers", "Authorization,content-type")
				w.Header().Set("Access-Control-Allow-Methods", "GET,PATCH,DELETE,OPTIONS")
			}
		},
	)

	api.AddResource(
		model.User{},
		resource.UserResource{
			UserStorage: userStorage,
		},
	)
	api.AddResource(
		model.Trip{},
		resource.TripResource{
			TripStorage:     tripStorage,
			UserStorage:     userStorage,
			CarShareStorage: carShareStorage,
			Clock:           clock.New(),
		},
	)
	api.AddResource(
		model.CarShare{},
		resource.CarShareResource{
			CarShareStorage: carShareStorage,
			TripStorage:     tripStorage,
			UserStorage:     userStorage,
			TokenVerifier:   tokenVerifier,
		},
	)

	// handler for metrics
	r.GET("/metrics", gin.WrapH(promhttp.Handler()))

	log.Infof("Listening and serving HTTP on :%d", *port)
	log.Fatal(r.Run(fmt.Sprintf(":%d", *port)))
}
